cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -710,12 +732,42 @@ public int deleteAllIndexes(int limit) {
     return count;
   }
 
-  // TODO: CDAP-5663 The entire logic of mapping between Indexers and keys should be made internal to MetadataDataset
-  private Indexer getIndexerForKey(String key) {
-    if ("schema".equals(key)) {
-      return new SchemaIndexer();
+  /**
+   * Returns all the {@link Indexer indexers} that apply to a specified metadata key.
+   *
+   * @param key the key for which {@link Indexer indexers} are required
+   */
+  private Set<Indexer> getIndexersForKey(String key) {
+    // only DefaultIndexer for user scope
+    if (MetadataScope.USER == scope) {
+      return Collections.<Indexer>singleton(new DefaultValueIndexer());
+    }
+    Set<Indexer> indexers = new HashSet<>();
+    switch (key) {
+      case AbstractSystemMetadataWriter.SCHEMA_KEY:
+        indexers.add(new SchemaIndexer());
+        break;
+      case AbstractSystemMetadataWriter.ENTITY_NAME_KEY:
+        // used for listing entities sorted in ascending order of name
+        indexers.add(new ValueOnlyIndexer());
```

Comment:
Thanks. Addressed comment about maintaining a map from key to indexers. 

Regarding the other comment, my follow-up PR (https://github.com/caskdata/cdap/compare/feature/more-indexes-contd...feature/search-rest-api?expand=1) adds a method to Indexer to get the sort order, and then a way to determine the column for the given key and sort order. Don't want to tie an indexer directly to an index column, because the same indexer can be used by multiple columns.